### PR TITLE
Initial ITs

### LINKED ISSUE #11: Update to use newer version of Nexus Repo
Thanks for creating an issue! Please fill out this form so we can be
sure to have all the information we need, and to minimize back and forth.

* **What are you trying to do?**
Right now this plugin is based on Nexus Repo 3.8.0-02, we should update it!

* **What feature or behavior is this required for?**
Just ensures compatibility with newer versions of Nexus Repo, as some dependencies have changed (commons-compress, etc...)

* **How could we solve this issue? (Not knowing is okay!)**
Update the `pom.xml`, `Dockerfile` and `README.md` to the new versions, and as well test these changes against a newer version of Nexus Repo, updating any API calls that might have changed or been deprecated along the way

* **Anything else?**
Have fun!

### PR DESCRIPTION
**Pull request by allenhsieh**
This PR sets up the structure for ITs and includes a few ITs

This pull request makes the following changes:
* Sets up IT Structure (testData folders/constants/etc)
* Starting code for ArtifactsXmlAbsoluteUrlRemover IT test
* Transitions from TravisCI to CircleCI

Fixes #11

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -1,102 +1,114 @@
...
+            <artifactId>nexus-staging-maven-plugin</artifactId>
+            <version>1.6.8</version>
+            <extensions>true</extensions>
+            <configuration>
+              <serverId>rso</serverId>
+              <nexusUrl>https://repository.sonatype.org/</nexusUrl>
+              <autoReleaseAfterClose>false</autoReleaseAfterClose>
+            </configuration>
+          </plugin>
+          <plugin>
```
[the remark this answers is not in the document]
Comment (allenhsieh) [reply pom.xml]: in: e294875

File: pom.xml [hunk 2]
```diff
@@ -1,102 +1,114 @@
...
+          <name>performRelease</name>
+          <value>true</value>
+        </property>
+      </activation>
+      <build>
+        <plugins>
+          <!--
+          Ensure use of Java 8, as required by NXRM
+          -->
+          <plugin>
```
[the remark this answers is not in the document]
Comment (allenhsieh) [reply pom.xml]: in: e294875

File: pom.xml [hunk 3]
```diff
@@ -1,102 +1,114 @@
...
+                <configuration>
+                  <rules>
+                    <requireJavaVersion>
+                      <version>[1.8,1.9)</version>
+                    </requireJavaVersion>
+                  </rules>
+                </configuration>
+              </execution>
+            </executions>
+          </plugin>
```
[the remark this answers is not in the document]
Comment (allenhsieh) [reply pom.xml]: in: https://github.com/sonatype-nexus-community/nexus-repository-p2/pull/17/commits/a664b9efbf3fda34f2a367dea5ff62c9f6916b37

File: nexus-repository-p2-it/src/test/it-resources/artifacts-mirror-removed.xml [hunk 4]
```diff
@@ -0,0 +1,49 @@
+<?xml version="1.0" encoding="UTF-8"?><?artifactRepository class='org.eclipse.equinox.internal.p2.artifact.repository.simple.SimpleArtifactRepository' version='1.0.0'?><repository name="HelloWorld Artifacts" type="org.eclipse.equinox.p2.artifact.repository.simpleRepository" version="1.0.0">
```
Comment (aornatovskyy) [inline nexus-repository-p2-it/src/test/it-resources/artifacts-mirror-removed.xml]: what this file for?

File: nexus-repository-p2-it/src/test/java/org/sonatype/nexus/repository/p2/internal/P2ProxyIT.java [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+import org.junit.Before;
+import org.junit.Test;
+import org.ops4j.pax.exam.Configuration;
+import org.ops4j.pax.exam.Option;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.is;
+import static org.sonatype.nexus.testsuite.testsupport.FormatClientSupport.status;
+
```
Comment (aornatovskyy) [inline nexus-repository-p2-it/src/test/java/org/sonatype/nexus/repository/p2/internal/P2ProxyIT.java]: looks like one extra line

File: nexus-repository-p2-it/src/test/java/org/sonatype/nexus/repository/p2/internal/P2ProxyIT.java [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+    try (CloseableHttpResponse response = proxyClient.get(ARTIFACT_XML)) {
+      HttpEntity entity = response.getEntity();
+      String result = IOUtils.toString(entity.getContent()).replaceAll("\\s+","");
+
+      InputStream targetStream = new FileInputStream(testData.resolveFile(ARTIFACT_WITHOUT_MIRROR_XML));
+      String expected = IOUtils.toString(targetStream).replaceAll("\\s+","");
+
+      assertThat(result, equalTo(expected));
+
```
Comment (aornatovskyy) [inline nexus-repository-p2-it/src/test/java/org/sonatype/nexus/repository/p2/internal/P2ProxyIT.java]: IMHO: extra line

File: nexus-repository-p2-it/src/test/java/org/sonatype/nexus/repository/p2/internal/fixtures/P2RepoRecipes.groovy [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+/**
+ * Factory for p2 {@link Repository} {@link Configuration}
+ */
+@CompileStatic
+trait P2RepoRecipes
+    extends ConfigurationRecipes
+{
+  @Nonnull
+  Repository createP2Proxy(final String name,
+                             final String remoteUrl)
```
Comment (aornatovskyy) [inline nexus-repository-p2-it/src/test/java/org/sonatype/nexus/repository/p2/internal/fixtures/P2RepoRecipes.groovy]: broken formating on line 33

File: nexus-repository-p2/src/test/resources/org/sonatype/nexus/repository/p2/internal/metadata/artifacts.xml [hunk 8]
```diff
@@ -0,0 +1,52 @@
+<?xml version='1.0' encoding='UTF-8'?>
```
Comment (aornatovskyy) [inline nexus-repository-p2/src/test/resources/org/sonatype/nexus/repository/p2/internal/metadata/artifacts.xml]: the same. what this file for?

File: pom.xml [hunk 9]
```diff
@@ -1,102 +1,128 @@
...
+    <!-- ensure we can find the parent pom when starting from an empty local .m2 repository -->
+    <repository>
+      <id>rso-snapshots</id>
+      <url>https://repository.sonatype.org/content/repositories/snapshots</url>
+    </repository>
+  </repositories>
 
   <build>
     <plugins>
+      <!--
```
Comment (aornatovskyy) [inline pom.xml]: is this comment is necessary? we should have in README prerequisites information about java. 
not sure it is good to keep such information on both places.

File: pom.xml [hunk 10]
```diff
@@ -1,102 +1,128 @@
...
+        <executions>
+          <execution>
+            <id>enforce-java</id>
+            <goals>
+              <goal>enforce</goal>
+            </goals>
+            <configuration>
+              <rules>
+                <requireJavaVersion>
+                  <version>[1.8,1.9)</version>
```
Comment (aornatovskyy) [inline pom.xml]: we had lots of problems with  java  > 8 version. I think it will be good to remove 1.9

**Comment by allenhsieh** [comment]
> @allenhsieh I would suggest for next formats to separate PRs for adding new IT tests with migration for new structure.

@aornatovskyy Sounds good! I will make sure to remember this for the future

File: nexus-repository-p2-it/src/test/it-resources/artifacts-mirror-removed.xml (hunk 4 above)
Comment (allenhsieh) [reply nexus-repository-p2-it/src/test/it-resources/artifacts-mirror-removed.xml]: I am unsure myself. I believe I took most of the structure from the Conda Community Format. I will check in with @DarthHater and @bhamail

File: nexus-repository-p2/src/test/resources/org/sonatype/nexus/repository/p2/internal/metadata/artifacts.xml (hunk 8 above)
Comment (allenhsieh) [reply nexus-repository-p2/src/test/resources/org/sonatype/nexus/repository/p2/internal/metadata/artifacts.xml]: @DarthHater @bhamail

File: pom.xml (hunk 9 above)
Comment (allenhsieh) [reply pom.xml]: @DarthHater @bhamail

File: pom.xml (hunk 10 above)
Comment (allenhsieh) [reply pom.xml]: @DarthHater @bhamail

File: nexus-repository-p2-it/src/test/java/org/sonatype/nexus/repository/p2/internal/fixtures/P2RepoRecipes.groovy (hunk 7 above)
Comment (allenhsieh) [reply nexus-repository-p2-it/src/test/java/org/sonatype/nexus/repository/p2/internal/fixtures/P2RepoRecipes.groovy]: in: https://github.com/sonatype-nexus-community/nexus-repository-p2/pull/17/commits/94912ebe5ebec6d38ce4e6c99a076d6a7b89a042

File: nexus-repository-p2-it/src/test/java/org/sonatype/nexus/repository/p2/internal/P2ProxyIT.java (hunk 5 above)
Comment (allenhsieh) [reply nexus-repository-p2-it/src/test/java/org/sonatype/nexus/repository/p2/internal/P2ProxyIT.java]: in: https://github.com/sonatype-nexus-community/nexus-repository-p2/pull/17/commits/94912ebe5ebec6d38ce4e6c99a076d6a7b89a042

File: nexus-repository-p2-it/src/test/java/org/sonatype/nexus/repository/p2/internal/P2ProxyIT.java (hunk 6 above)
Comment (allenhsieh) [reply nexus-repository-p2-it/src/test/java/org/sonatype/nexus/repository/p2/internal/P2ProxyIT.java]: in: https://github.com/sonatype-nexus-community/nexus-repository-p2/pull/17/commits/94912ebe5ebec6d38ce4e6c99a076d6a7b89a042

**Comment by DarthHater** [comment]
@aornatovskyy @allenhsieh I think that for when we setup a project with ITs for the first time, we should have at least 1-2 tests at the beginning, not just bare project setup. We want to ensure it's working, so maybe just limit it to a couple tests that go through some project specific files, and then stuff like what we did with the artifactsXML test can be done in follow up PRs, that sound ok?

File: pom.xml (hunk 10 above)
Comment (DarthHater) [reply pom.xml]: I had wondered the same thing. I think @aornatovskyy is right, limiting to 1.8 is fine!

File: nexus-repository-p2/src/test/resources/org/sonatype/nexus/repository/p2/internal/metadata/artifacts.xml (hunk 8 above)
Comment (DarthHater) [reply nexus-repository-p2/src/test/resources/org/sonatype/nexus/repository/p2/internal/metadata/artifacts.xml]: I believe this is for a Unit Test (don't quote me on that), where we test for if the artifacts.xml file is changed correctly.

File: nexus-repository-p2-it/src/test/it-resources/artifacts-mirror-removed.xml (hunk 4 above)
Comment (DarthHater) [reply nexus-repository-p2-it/src/test/it-resources/artifacts-mirror-removed.xml]: Pretty sure this is for a Unit Test that we repurposed for a IT (because what we tested is more of an integration test, but we wrote a pretty complicated Unit Test in the past to make up for there being no IT framework).

File: pom.xml (hunk 10 above)
Comment (allenhsieh) [reply pom.xml]: @aornatovskyy @DarthHater after some research, I believe Dan had it correct here. [This](https://maven.apache.org/enforcer/enforcer-rules/versionRanges.html) seems to suggest the way Dan has it set up, it will only cover all versions of 1.8 (i.e. equal and greater than 1.8 but LESS THAN 1.9)

File: nexus-repository-p2-it/src/test/java/org/sonatype/nexus/repository/p2/internal/P2ProxyIT.java [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+    assertThat(asset.contentType(), is(equalTo(MIME_TYPE)));
+    assertThat(asset.format(), is(equalTo(FORMAT_NAME)));
+
+    final Component component = findComponent(proxyRepo, COMPONENT_NAME);
+    assertThat(component.version(), is(equalTo(VERSION_NUMBER)));
+    assertThat(component.group(), is(equalTo(null)));
+  }
+
+  @Test
+  public void retrieveP2IndexFromProxyWhenRemoteOnline() throws Exception {
```
Comment (DarthHater) [inline nexus-repository-p2-it/src/test/java/org/sonatype/nexus/repository/p2/internal/P2ProxyIT.java]: @aornatovskyy we added these so we could do some refactoring in another PR, we discovered a bug that is fairly nasty and decided these tests would help us figure it out quicker

File: nexus-repository-p2-it/src/test/java/org/sonatype/nexus/repository/p2/internal/P2ProxyIT.java [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+
+    final Asset asset = findAsset(proxyRepo, P2_INDEX);
+    assertThat(asset.name(), is(equalTo(P2_INDEX)));
+    assertThat(asset.format(), is(equalTo(FORMAT_NAME)));
+    // TODO: We discovered these files do exist, but right now it's not working, we need to fix this!
+    // assertThat(status(proxyClient.get("folder/" + P2_INDEX)), is(HttpStatus.OK));
+  }
+
+  @Test
+  public void retrieveCompositeArtifactsJarFromProxyWhenRemoteOnline() throws Exception {
```
Comment (DarthHater) [inline nexus-repository-p2-it/src/test/java/org/sonatype/nexus/repository/p2/internal/P2ProxyIT.java]: @aornatovskyy we added these so we could do some refactoring in another PR, we discovered a bug that is fairly nasty and decided these tests would help us figure it out quicker

File: nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/internal/proxy/P2ProxyRecipe.groovy [hunk 13]
```diff
@@ -216,7 +216,7 @@ class P2ProxyRecipe
   }
 
   static TokenMatcher tokenMatcherForExtensionAndName(final String extension, final String name = '.+', final String path = '.+') {
-    new TokenMatcher("{path:.*}/{name:${name}}.{extension:${extension}}")
+    new TokenMatcher("{path:${path}}/{name:${name}}.{extension:${extension}}")
```
Comment (DarthHater) [inline nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/internal/proxy/P2ProxyRecipe.groovy]: Path was never being used, so we actually use it now. ITs check out with this change, and I'm fine with it sneaking through.

**Comment by bhamail** [comment]
I will fix the outdated travis build requirement...stand by...

**Comment by bhamail** [comment]
OK, CI build requirements should be fixed now, as well as deploy key. Merge away!
